src/com/cft/contactmerge/io/XmlImporter.java

```diff
@@ -1,31 +1,132 @@
+    private NodeList dataNodes;
+
+    public void Load(String filename) throws FileNotFoundException
     {
-        // TODO: Add code
+        if (filename == null || filename.trim() == "")
+        {
+            throw new IllegalArgumentException("filename is required");
+        }
+
+        File inputFile = new File(filename);
+
+        if (!inputFile.exists() || inputFile.isDirectory())
+        {
+            throw new FileNotFoundException(String.format("Unable to find %s", filename));
+        }
+
+        InputStream inputStream = new FileInputStream(inputFile);
+
+        Load(inputStream);
+    }
+
+    public void Load(InputStream inputStream)
+    {
+        if (inputStream == null)
+        {
+            throw new IllegalArgumentException("inputStream is required");
+        }
+
+        try {
+            if (!(inputStream.available() > 0))
+                throw new UnknownFormatConversionException("inputStream not available");
+        }
+        catch (IOException e)
+        {
+            throw new UnknownFormatConversionException("inputStream not available");
+        }
+
+        try {
+            DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
+            DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
+            Document doc = dBuilder.parse(inputStream);
+
+            columnMap = getXmlAttribuateMap(doc.getElementsByTagName("Field"));
+            dataNodes = doc.getElementsByTagName("row");
+        }
+        catch(SAXException | IOException | ParserConfigurationException e)
+        {
+            throw new UnknownFormatConversionException(e.getMessage());
+        }
+    }
+
+    // TODO: Need to figure out a flexible way to map XML values with
+    // properties. This is a temporary solution so that we can start playing
+    // with data from GiftWorks.
+    private Map<String, String> getXmlAttribuateMap(NodeList nodes)
+    {
+        Map<String, String> map = new HashMap<String, String>();
+
+        for (int i=0; i < nodes.getLength(); i++)
+        {
+            NamedNodeMap mapItem = nodes.item(i).getAttributes();
+            map.put(mapItem.item(1).getNodeValue(), mapItem.item(0).getNodeValue());
+        }
+
+        return map;
+    }
+
+    // TODO: Need to figure out a flexible way to map XML values with
+    // properties. This is a temporary solution so that we can start playing
+    // with data from GiftWorks.
+    private Map<String, String> getContactValues(Node node)
+    {
+        Map<String, String> values = new HashMap<String, String>();
+
+        for (int j=0; j < node.getAttributes().getLength(); j++) {
+            values.put(node.getAttributes().item(j).getNodeName(), node.getAttributes().item(j).getNodeValue());
+        }
+
+        return values;
     }
 
     public Iterator<Contact> iterator()
     {
+        if (dataNodes == null)
+        {
+            throw new IllegalStateException("Must load file before using the iterator");
```

Comment:
Something to think about is if we have different interfaces (web, gui, console)... how will they interpret the way these error messages should be shown.

Reply:
Yep. At some point we need to have a conversation on error handling.